Support list of list of relationships/nodes mapping (#2204)
spring-projects/spring-data-neo4j: src/main/asciidoc/appendix/custom-queries.adoc

.Example for 'The Matrix' and 'Keanu Reeves'
image::movie-graph-deep.png[]

[[custom-queries.for-relationships.long-paths.database]]
==== Database-side reduction

Keeping in mind that Spring Data Neo4j can only properly process record based, the result for one entity instance needs to be in one record.
Using https://neo4j.com/docs/cypher-manual/current/syntax/patterns/#cypher-pattern-path-variables[Cypher's path] capabilities is a valid option to fetch all branches in the graph.

[source,cypher]
.Naive path-based approach
----
MATCH p=(m:Movie{title: 'The Matrix'})<-[:ACTED_IN]-(:Person)-[:ACTED_IN*..0]->(:Movie)
RETURN p;
----

This will result in multiple paths that are not merged within one record.
It is possible to call `collect(p)` but Spring Data Neo4j does not understand the concept of paths in the mapping process.
Thus, nodes and relationships needs to get extracted for the result.

[source,cypher]
.Extracting nodes and relationships
----
MATCH p=(m:Movie{title: 'The Matrix'})<-[:ACTED_IN]-(:Person)-[:ACTED_IN*..0]->(:Movie)
RETURN m, nodes(p), relationships(p);
----

> Because there are multiple paths that lead from 'The Matrix' to another movie, the result will be still not a single record.

Reviewer: ```suggestion
Because there are multiple paths that lead from 'The Matrix' to another movie, the result still won't be a single record.
```

This is where https://neo4j.com/docs/cypher-manual/current/functions/list/#functions-reduce[Cypher's reduce function] comes into play.

[source,cypher]
.Reducing nodes and relationships
----
MATCH p=(m:Movie{title: 'The Matrix'})<-[:ACTED_IN]-(:Person)-[:ACTED_IN*..0]->(:Movie)
WITH collect(p) as paths, m
WITH m,
reduce(a=[], node in reduce(b=[], c in [aa in paths | nodes(aa)] | b + c) | case when node in a then a else a + node end) as nodes,
reduce(d=[], relationship in reduce(e=[], f in [dd in paths | relationships(dd)] | e + f) | case when relationship in d then d else d + relationship end) as relationships
RETURN m, relationships, nodes;
----

> It looks a little complicated at first sight, but the `reduce` function allows us to flatten the nodes and relationships from various paths.

Reviewer: ```suggestion
The `reduce` function allows us to flatten the nodes and relationships from various paths.
```
